nixos/modules/services/web-apps/otterwiki.nix

```diff
@@ -0,0 +1,167 @@
+  secretKeyFile = inst: "/var/lib/${inst.process.unit}/secret.cfg";
+
+  gunicornInstanceOptions =
+    { config, ... }@moduleArgs:
+    lib.recursiveUpdate (import ../web-servers/gunicorn/instance-options.nix moduleArgs) {
+      options.process.unit.default = "otterwiki-${config._module.args.name}";
+    };
+
+  instanceOptions =
+    { config, ... }@moduleArgs:
+    {
+      options = {
+        inherit ((gunicornInstanceOptions moduleArgs).options) process socket;
+
+        package = lib.mkPackageOption pkgs "otterwiki" { };
+
+        repository = lib.mkOption {
+          type = lib.types.str;
+          default = "/var/lib/${config.process.unit}/repository";
+          description = "Path to the Git repository for the wiki content.";
+        };
+
+        databaseUri = lib.mkOption {
+          type = lib.types.str;
+          default = "sqlite:////var/lib/${config.process.unit}/db.sqlite";
+          description = ''
+            SQLAlchemy database connection URI.
+            Defaults to using an independent SQLite database.
+          '';
+        };
+
+        settings = lib.mkOption {
+          type = lib.types.attrsOf lib.types.str;
+          default = { };
+          example = {
+            SITE_NAME = "Otterwiki";
+          };
+          description = ''
+            Configuration for Otter Wiki.
+            See <https://otterwiki.com/Configuration>.
+
+            A cookie secret is automatically generated at first start.
+          '';
+        };
+
+        environmentFile = lib.mkOption {
+          type = lib.types.nullOr lib.types.path;
+          default = null;
+          description = ''
+            Path to an environment file.
+            This is suitable for supplying secrets such as the SMTP credentials.
+          '';
+        };
+      };
+    };
+
+in
+{
+  options.services.otterwiki.instances = lib.mkOption {
+    type = lib.types.attrsOf (lib.types.submodule instanceOptions);
+    default = { };
+    description = "Configuration for instances of Otter Wiki.";
+  };
+
+  config = {
+    services.gunicorn.instances = mapToUnits (inst: {
+      inherit (inst) process socket;
+      app.package = inst.package;
+    }) cfg.instances;
+
+    systemd.services = mapToUnits (inst: {
+      environment = {
+        REPOSITORY = inst.repository;
+        SQLALCHEMY_DATABASE_URI = inst.databaseUri;
+        OTTERWIKI_SETTINGS = secretKeyFile inst;
+      }
+      // inst.settings;
+
+      path = [ pkgs.git ];
+
+      preStart = ''
+        # generate cookie secret
+        if [[ ! -f ${lib.escapeShellArg (secretKeyFile inst)} ]]; then
+          (
+            echo -n "SECRET_KEY='"
+            cat /dev/urandom | head -c128 | tr -dc A-Za-z0-9
+            echo "'"
+          ) > ${lib.escapeShellArg (secretKeyFile inst)}
+        fi
+
+        # init wiki repo
+        if [[ ! -d ${lib.escapeShellArg inst.repository} ]]; then
```

Comment:
Consider adding an `mkdir -p $(dirname <file>)` for this. Alternatively, bindmount `repository` into `/var/lib/${unit}` and just assume the location for the rest of the service.